File: kraken-engine/src/main/java/kraken/runtime/engine/handlers/NumberSetPayloadHandler.java
/*
 *  Copyright 2023 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package kraken.runtime.engine.handlers;

import java.util.List;

import javax.money.MonetaryAmount;

import kraken.el.functionregistry.functions.MoneyFunctions;
import kraken.el.math.Numbers;
import kraken.model.payload.PayloadType;
import kraken.runtime.EvaluationSession;
import kraken.runtime.engine.RulePayloadHandler;
import kraken.runtime.engine.context.data.DataContext;
import kraken.runtime.engine.handlers.trace.FieldValueValidationOperation;
import kraken.runtime.engine.result.NumberSetPayloadResult;
import kraken.runtime.engine.result.PayloadResult;
import kraken.runtime.expressions.KrakenExpressionEvaluator;
import kraken.runtime.model.rule.RuntimeRule;
import kraken.runtime.model.rule.payload.validation.NumberSetPayload;
import kraken.tracer.Tracer;

/**
 * @author Mindaugas Ulevicius
 */
public class NumberSetPayloadHandler implements RulePayloadHandler {
    private final KrakenExpressionEvaluator evaluator;

    public NumberSetPayloadHandler(KrakenExpressionEvaluator evaluator) {
        this.evaluator = evaluator;
    }

    @Override
    public PayloadType handlesPayloadType() {
        return PayloadType.NUMBER_SET;
    }

    @Override
    public PayloadResult executePayload(RuntimeRule rule, DataContext dataContext, EvaluationSession session) {
        var payload = (NumberSetPayload) rule.getPayload();

        var value = evaluator.evaluateTargetField(rule.getTargetPath(), dataContext);
        boolean success = true;
        if(value instanceof MonetaryAmount) {
            value = MoneyFunctions.fromMoney((MonetaryAmount) value);
        }
        if(value instanceof Number) {
            Tracer.doOperation(new FieldValueValidationOperation(value));
            success = Numbers.isValueInNumberSet((Number) value, payload.getMin(), payload.getMax(), payload.getStep());
        }

        var templateVariables = evaluator.evaluateTemplateVariables(payload.getErrorMessage(), dataContext, session);
        return new NumberSetPayloadResult(success, payload, templateVariables);
    }

    @Override
    public String describePayloadResult(PayloadResult payloadResult) {
        var result = (NumberSetPayloadResult) payloadResult;
        return result.getSuccess()
            ? String.format("Field is valid. Field value is in %s.", describeExpectedNumberSet(result))
            : String.format("Field is not valid. Field value is not in %s.", describeExpectedNumberSet(result));
    }

    private String describeExpectedNumberSet(NumberSetPayloadResult payloadResult) {
        return String.format(
            "number set [%s, %s]%s",
            payloadResult.getMin() != null ? payloadResult.getMin() : "-∞",
            payloadResult.getMax() != null ? payloadResult.getMax() : "∞",
            payloadResult.getStep() != null ? " with step " + payloadResult.getStep() : ""
        );
    }
}

File: javascript/packages/engine/src/engine/handlers/NumberSetPayloadHandler.ts
/*
 *  Copyright 2023 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Payloads, Rule } from 'kraken-model'
import PayloadType = Payloads.PayloadType
import { RulePayloadHandler } from './RulePayloadHandler'
import { ExpressionEvaluator } from '../runtime/expressions/ExpressionEvaluator'
import { NumberSetPayloadResult } from 'kraken-engine-api'
import { ExecutionSession } from '../ExecutionSession'
import { DataContext } from '../contexts/data/DataContext'
import { payloadResultCreator } from '../results/PayloadResultCreator'
import NumberSetPayload = Payloads.Validation.NumberSetPayload
import { Numbers } from '../runtime/expressions/math/Numbers'
import { Moneys } from '../runtime/expressions/math/Moneys'
import { logger } from '../../utils/DevelopmentLogger'

export class NumberSetPayloadHandler implements RulePayloadHandler {
    constructor(private readonly evaluator: ExpressionEvaluator) {}

    handlesPayloadType(): PayloadType {
        return PayloadType.NUMBER_SET
    }
    executePayload(rule: Rule, dataContext: DataContext, session: ExecutionSession): NumberSetPayloadResult {
        const payload = rule.payload as NumberSetPayload

        let value = this.evaluator.evaluateTargetField(rule.targetPath, dataContext)
        if (Moneys.isMoney(value)) {
            value = value.amount
        }

        let success = true
        if (typeof value === 'number') {
            logger.debug(() => `Validating field which has value: ${ExpressionEvaluator.renderFieldValue(value)}`)
            success = Numbers.isValueInNumberSet(value, payload.min, payload.max, payload.step)
        }

        const templateVariables = this.evaluator.evaluateTemplateVariables(payload.errorMessage, dataContext, session)

        return payloadResultCreator.numberSet(payload, success, templateVariables)
    }

    describePayloadResult(payloadResult: NumberSetPayloadResult): string {
        const min = payloadResult.min !== undefined ? payloadResult.min : '-∞'
        const max = payloadResult.max !== undefined ? payloadResult.max : '∞'
        const stepMsg = payloadResult.step !== undefined ? ` with step ${payloadResult.step}` : ''
        const numberSetDescription = `number set [${min}, ${max}]${stepMsg}`
        return payloadResult.success
            ? `Field is valid. Field value is in ${numberSetDescription}.`
            : `Field is not valid. Field value is not in ${numberSetDescription}.`
    }
}
